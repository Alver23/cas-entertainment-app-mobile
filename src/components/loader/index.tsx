// Dependencies
import React, { ReactElement } from 'react';
import { ActivityIndicator, StatusBar, View } from 'react-native';

// Hooks
import { useTheme } from '@hooks/theme';
// Styles
import useStyles from './styles';

const Component = (): ReactElement => {
  const styles = useStyles();
  const theme = useTheme();
  return (
    <View style={[styles.container, styles.horizontal]}>
      <ActivityIndicator size={'large'} color={theme.colors.primary} />
      <StatusBar barStyle="dark-content" backgroundColor={theme.colors.white} />
    </View>
  );
};

export const LoaderView = React.memo(Component);
