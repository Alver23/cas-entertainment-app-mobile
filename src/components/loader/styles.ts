// Dependencies
import { StyleSheet } from 'react-native';

// Hooks
import { makeStyles } from '@hooks/theme';

const styles = makeStyles(({ colors }) => ({
  container: {
    ...StyleSheet.absoluteFillObject,
    flex: 1,
    justifyContent: 'center',
    backgroundColor: colors.white,
    alignItems: 'center',
    zIndex: 3,
  },
  horizontal: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    padding: 10,
  },
}));

export default styles;
