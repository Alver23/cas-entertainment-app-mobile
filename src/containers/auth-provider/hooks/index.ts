// Dependencies
import React from 'react';

// Hooks
import { useDidMount } from '@hooks/did-mount';

// Models
import { IMutableRefObject } from '@models/mutable-ref-interface';

// Commons
import { clearAuthData, getToken } from '@commons/auth';

// Contexts
import { IAuthContext } from '@contexts/auth';

export const useAuth = (): [T, () => void] => {
  const initialState: IAuthContext = {
    token: null,
    authenticated: false,
    initialize: false,
  };

  const [data, setData] = React.useState(initialState);
  const isMounted: IMutableRefObject<boolean> = React.useRef(true);

  const logout: () => void = React.useCallback(async () => {
    setData({ ...initialState, initialize: true });
    await clearAuthData();
  }, []);

  useDidMount(
    (): Function => {
      return (): void => {
        isMounted.current = false;
      };
    },
  );

  useDidMount((): void => {
    getToken()
      .then((token: string) => {
        if (isMounted.current) {
          setData({
            token,
            initialize: true,
            authenticated: !!token,
          });
        }
      })
      .catch(() => {
        if (isMounted.current) {
          setData(initialState);
        }
      });
  });

  return [data, logout];
};
