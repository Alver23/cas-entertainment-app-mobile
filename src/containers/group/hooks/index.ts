// Hooks
import { useFetch } from '@hooks/use-fetch';

// Services
import { groupService } from '@services/group';

export const useGroupList = (groupIds: number[]) => {
  const initialState = {
    data: { items: [] },
  };

  const [
    {
      data: { items },
      loading,
    },
  ] = useFetch(
    () => {
      const searchParams = new URLSearchParams();
      groupIds.forEach((item) => {
        searchParams.append('groupId', (item as unknown) as string);
      });

      return groupService.get(searchParams);
    },
    [groupIds],
    initialState,
  );

  return [items, loading];
};
