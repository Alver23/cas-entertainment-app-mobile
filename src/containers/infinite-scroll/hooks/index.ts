// Dependencies
import React from 'react';

// Hooks
import { useParamList } from '@hooks/params';

// Reducer
import { initialState, reducer, ACTIONS, IState } from '../reducer';

export const useInfiniteScroll = (fetcher: any): [IState, () => void, () => void] => {
  const [, setItems] = React.useState<Array<any>>([]);
  const [state, dispatch] = React.useReducer(reducer, initialState);

  const isMounted = React.useRef<boolean>(true);
  const hasNextPageRef = React.useRef<boolean>(true);
  const [params, setParams, , initialParams] = useParamList();
  const { name: filterName } = params;

  const updateItems = (newItems): void => {
    setItems((prevState) => {
      const buildItems = prevState.concat(newItems);
      dispatch({ type: ACTIONS.SET_DATA, payload: buildItems });
      return buildItems;
    });
  };

  const fetchData = (parameters): Promise<boolean> => {
    return new Promise((resolve) => {
      fetcher(parameters)
        .then((response) => {
          if (isMounted.current) {
            const { items: newItems, totalPages, totalItems: total } = response;
            const nextPage = parameters.page + 1;
            const hasPage = parameters.page !== totalPages;
            dispatch({ type: ACTIONS.SET_TOTAL_ITEMS, payload: total });
            hasNextPageRef.current = hasPage;
            updateItems(newItems);
            setParams((prevState) => ({
              ...prevState,
              page: nextPage,
            }));
            resolve(false);
          }
        })
        .catch((error: Error) => {
          if (isMounted.current) {
            dispatch({ type: ACTIONS.SET_ERROR, payload: error });
          }
          resolve(false);
        });
    });
  };

  const handleLoadMore = React.useCallback((): void => {
    if (hasNextPageRef.current) {
      dispatch({ type: ACTIONS.SET_LOADER, property: 'loading', payload: true });
      fetchData(params).finally(() => {
        dispatch({ type: ACTIONS.SET_LOADER, property: 'loading', payload: false });
      });
    }
  }, [params]);

  const initialFetch = React.useCallback((propertyAction = 'refreshing', newParams = {}) => {
    setItems([]);
    dispatch({ type: ACTIONS.SET_LOADER, property: propertyAction, payload: true });
    const buildParams = { ...params, ...initialParams, ...newParams };
    setParams(buildParams);
    return fetchData(buildParams).finally(() => {
      dispatch({ type: ACTIONS.SET_LOADER, property: propertyAction, payload: false });
    });
  }, []);

  React.useEffect(
    (): (() => void) => (): void => {
      isMounted.current = false;
    },
    [],
  );

  React.useEffect(() => {
    (async () => {
      dispatch({ type: ACTIONS.SET_LOADER, property: 'initLoading', payload: true });
      await initialFetch();
      dispatch({ type: ACTIONS.SET_LOADER, property: 'initLoading', payload: false });
    })();
  }, []);

  React.useEffect(() => {
    (async () => {
      if (filterName) {
        await initialFetch('initLoading', { name: filterName });
      }
    })();
  }, [filterName]);

  return [state, initialFetch, handleLoadMore];
};
