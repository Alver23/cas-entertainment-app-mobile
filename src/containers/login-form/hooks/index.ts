// Dependencies
import React from 'react';
import RNRestart from "react-native-restart";

// Hooks
import { useShowAppLoader } from '@hooks/app-loader';
import { useStateAction } from '@hooks/state-action';

// Services
import { authenticationService } from '@services/authentication/authentication-service';

// Commons
import { setAuthData } from "@commons/auth";

export const useAuthentication = (): [boolean, (values) => void] => {
  const [showAppLoader] = useShowAppLoader();
  const [error, showError, hideError] = useStateAction();

  const onSubmit = React.useCallback((values) => {
    showAppLoader(true);
    authenticationService
      .login(values)
      .then(async ({ data }) => {
        await setAuthData(data);
        hideError();
        RNRestart.Restart();
      })
      .catch(() => {
        showError()
      })
      .finally(() => showAppLoader(false))
  }, []);

  return [error, onSubmit]
}
