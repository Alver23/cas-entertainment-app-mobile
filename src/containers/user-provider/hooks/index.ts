// Dependencies
import React from 'react';
import jwtDecode from 'jwt-decode';

// Hooks
import { useLocalSession } from '@hooks/auth';

// Services
import { userService } from '@services/user';

// Reducers
import { USER_CONSTANTS, initialState, userReducer, IState } from "../reducers";

export const useUserProvider = (): [IState] => {
  const { token } = useLocalSession();
  const [ state, dispatch ] = React.useReducer(userReducer, initialState);
  const { user } = state;

  React.useEffect(() => {
    const { email, firstName, lastName, roles, sub } = jwtDecode(token);
    const payload = { email, firstName, lastName, roles, id: sub };
    dispatch({ type: USER_CONSTANTS.SET_USER, payload });
  }, [token]);

  React.useEffect(() => {
    if (user?.id) {
      dispatch({ type: USER_CONSTANTS.SET_LOADING, payload: true });
      userService
        .getMenus(user.id)
        .then((payload) => {
          dispatch({ type: USER_CONSTANTS.SET_MENUS, payload });
        })
        .catch((error) =>  {
          dispatch({ type: USER_CONSTANTS.SET_ERROR, payload: error });
        })
        .finally(() => {
          dispatch({ type: USER_CONSTANTS.SET_LOADING, payload: false });
        });
    }
  }, [user?.id]);

  return [state];
}
