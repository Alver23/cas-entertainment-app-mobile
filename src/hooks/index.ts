export { useTheme, makeStyles } from './theme';
