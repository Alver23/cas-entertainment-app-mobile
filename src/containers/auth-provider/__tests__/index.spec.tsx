// Dependencies
import React, { PropsWithChildren, ReactElement } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

// Components
import { LoaderView } from '@components/loader';

// Hooks
import * as hooks from '../hooks';

// Components
import { AuthProvider } from '../index';

describe('AuthProvider', () => {
  let component: ShallowWrapper<PropsWithChildren<any>>;

  const FakeComponent = (): ReactElement => <></>;

  beforeEach(() => {
    component = shallow(
      <AuthProvider>
        <FakeComponent />
      </AuthProvider>,
    );
  });

  it('should render correctly when not initialize authentication', () => {
    expect(component.find(LoaderView).exists()).toBeTruthy();
  });

  it('should save a snapshot of the component', () => {
    const mockState: [Record<string, boolean>, () => void] = [{ initialize: true }, jest.fn()];
    jest.spyOn(hooks, 'useAuth').mockReturnValue(mockState);
    component.setProps({});
    const {
      children: { type },
    } = component.props();
    expect(type()).toBeTruthy();
    expect(component).toMatchSnapshot();
  });
});
