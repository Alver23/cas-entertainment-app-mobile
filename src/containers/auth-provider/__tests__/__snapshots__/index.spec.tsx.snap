// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthProvider should save a snapshot of the component 1`] = `
<ContextProvider
  value={
    Object {
      "initialize": true,
      "logout": [MockFunction],
    }
  }
>
  <FakeComponent />
</ContextProvider>
`;
